### DISCUSSION TITLE
M5StackでSDカードを有効にするとWiFiClientSecureが動かなくなる

### DISCUSSION THREAD

**Discussion by sesgak** [1ft-seabass/ma-hackathon-qa-and-knowledge-202210#14] [Q&A] [answered]
[esp32_CloudSpeech](https://github.com/MhageGH/esp32_CloudSpeech)
を利用してGoogleCloudのSTTで音声認識を実装しています。
音声認識単体ではうまく動き、こちらの発話を文字起こしすることができました。

続いて、別のスケッチでSDカードに読み込ませた画像を表示する機能を実装しました。
画像表示単体はうまく動き、画像を表示することができました。

その音声認識機能とSDカード認識機能を１つのスケッチに纏めて同時に動かそうとした際に、音声認識が動かなくなりました。

`M5.begin(true, false, true);`
でSDカード読み込みをfalseにすると動きますが、

`M5.begin(true, true, true);`
or
`M5.begin();`
だと動かなくなります。

WiFiClientSecureで接続する箇所でConnection failedになってしまいます。

```
const char*  server = "speech.googleapis.com";
const char* root_ca= 
    "-----BEGIN CERTIFICATE-----\n"
　省略
    "-----END CERTIFICATE-----\n";

CloudSpeechClient::CloudSpeechClient(Authentication authentication) {
  this->authentication = authentication;
  client.setCACert(root_ca);
  if (!client.connect(server, 443)){
    Serial.println("Connection failed!");
  }
}

```

何か理由や対処をご存じでしたら教えていただきたいです。

**Comment by 1ft-seabass** [comment, marked as answer]
今日はもう寝てしまうので、即答は難しいかもですが。まず拝見します！

**Reply by 1ft-seabass** [reply]
ちょっと私は経験ないんですが、まず M5.begin のパラメータ見てます。
https://qiita.com/penguinprogrammer/items/5abf3e4f9583e41e9cfa

**Reply by 1ft-seabass** [reply]
うーん、ちょっと指定パラメータがおかしいって感じじゃなさそうですね。

**Reply by 1ft-seabass** [reply]
@sesgak シリアルモニタで詳細ログとか分かりますかね？つながらないときに、何かが起きてるかもしれません。

ちょっと怪しんでいるのは、SD カードで JPEG 読む機能などはメモリや処理を結構消費しがちなので、おなじく負荷がかかりやすい Wi-Fi 系の初動がバッティングして、リソースを食い合って、不安定になっているということが考えられます。

そういう意味でシリアルモニタでログを追うと、より状況が見えるかもしれないです。

**Reply by 1ft-seabass** [reply]
対症療法にはなるのですが、SD カードの処理と Wi-Fi の処理を delay などで離してみるとかも効くかもしれません～。

**Reply by 1ft-seabass** [reply]
ひきつづき、しらべてみます。

**Reply by sesgak** [reply]
Serial.begin(115200);
削除しました。シリアルモニタは問題なく見れました！
エラーはまだ継続です。

**Reply by 1ft-seabass** [reply]
@sesgak おお！検証ありがとうございます～。SD と Wi-Fi（特にSecureがらみ） の相性かもしれないので調べてみます。

**Reply by 1ft-seabass** [reply]
@sesgak めちゃくちゃ余談情報なんですが、WiFiClientSecure が急につながらなくなるというケースはあって原因不明なので HTTPClient にしたら問題なくなったという話は私も体験しちゃので、もしかすると、WiFiClientSecure を HTTPClient に置き換えるとうまくいきかもしれません。おそらく、証明書ファイル系で不都合があって急に不通になるってかんじですね。

WiFiClientSecure でなく HTTPClient を使って M5Stack から HTTPS で AirTable にデータを書き込むメモ – 1ft-seabass.jp.MEMO
https://www.1ft-seabass.jp/memo/2022/05/14/m5stack-airtable-api-connect-using-httpclient/

**Reply by 1ft-seabass** [reply]
@sesgak その上で元ソースの https://github.com/MhageGH/esp32_CloudSpeech/tree/master/esp32_CloudSpeech みてみます。

**Reply by 1ft-seabass** [reply]
@sesgak ちょっと色々調べてるんですが、明確に効きそうなナレッジがまだない状況です。途中経過ですがお伝えしておきますー。

**Reply by sesgak** [reply]
ありがとうございます！

**Reply by sesgak** [reply]
https://tomosoft.jp/design/?p=45300

SDカード読み込みは諦めて、リンクを参考にWeb上から画像を取得するようにしました！

**Reply by sesgak** [reply]
次はaquestalkとesp32_CloudSpeechのI2Sの設定が競合したのか、esp32_CloudSpeech後にaquestalkを実行するとエラーに…
それぞれで色々設定しているので何かがおかしくなっているようです…

エラー
`E (93762) I2S: /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/driver/i2s.c:1187 (i2s_write):tx NULL`

**Reply by sesgak** [reply]
aquestalkの出力部分です。
```
int DAC_Write(int len, int16_t *wav){
  int i;

  for (i = 0; i < len; i++) {
    // upsampling x3
    int16_t wav3[3];
    AqResample_Conv(wav[i], wav3);
    // write to I2S DMA buffer
    for (int k = 0; k < 3; k++) {
      size_t transBytes;
      uint16_t sample[2];
      uint16_t us = ((uint16_t)wav3[k]) ^ 0x8000U;  // signed -> unsigned data 内蔵DA Only
      sample[0] = sample[1] = us; // mono -> stereo
      int iret = i2s_write((i2s_port_t)i2s_num, (const char*)sample, 4, &transBytes, portMAX_DELAY);
      if (iret < 0) {
        return iret; // -1:ESP_FAIL
      }
      if (iret == 0) {
        break;  //  0:TIMEOUT
      }
    }
  }
  return i;
}
```

↑のコードのここでエラーになりました。
`      int iret = i2s_write((i2s_port_t)i2s_num, (const char*)sample, 4, &transBytes, portMAX_DELAY);`

m5stack起動後に①aquestalkを実行した場合は正常に動きますが、esp32_CloudSpeech後に②aquestalkを実行すると上記の箇所でエラーになります。

①、②どちらも i2s_writeのパラメータは一緒だったので、謎は深まるばかりです。

**Reply by 1ft-seabass** [reply]
うおお、解決！とおもったら・・・。

i2s_writeまわりはハードウェアまわりのコードなだけに、なぜ画像ロードが影響しちゃうのか謎ですねー。

**Reply by sesgak** [reply]
画像ロードはWeb画像表示でうまくいったのですが、aquestalkでエラーになっちゃう感じです。
もうここはあきらめます！

**Reply by sesgak** [reply]
色々調べていただいてありがとうございました！

**Reply by 1ft-seabass** [reply]
チャレンジお疲れ様でした！力及ばずですいません！

引き続き、制作物応援しております！
